tmt/steps/report/junit.py

```diff
@@ -58,40 +42,170 @@ def duration_to_seconds(duration: Optional[str]) -> Optional[int]:
+    :param template_path: if set, the provided template will be used instead of
+        a pre-defined flavor template. In this case, the ``flavor`` must be set
+        to ``custom`` value.
+    :pawam extra_variables: if set, these variables get propagated into the
+        Jinja template.
+    """
+
+    # Get the template context for TMT results
+    results_context = results_context or ResultsContext(report.step.plan.execute.results())
 
-    for result in report.step.plan.execute.results():
+    # Prepare the template environment
+    def _read_log(log: Path) -> str:
+        """ Read the contents of a given result log """
         try:
-            main_log = report.step.plan.execute.read(result.log[0])
-        except (IndexError, AttributeError):
-            main_log = None
-        case = junit_xml.TestCase(
-            result.name,
-            classname=None,
-            elapsed_sec=duration_to_seconds(result.duration))
-
-        if report.data.include_output_log:
-            case.stdout = main_log
-
-        # Map tmt OUTCOME to JUnit states
-        if result.result == tmt.result.ResultOutcome.ERROR:
-            case.add_error_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.FAIL:
-            case.add_failure_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.INFO:
-            case.add_skipped_info(result.result.value, output=result.failures(main_log))
-        elif result.result == tmt.result.ResultOutcome.WARN:
-            case.add_error_info(result.result.value, output=result.failures(main_log))
-        # Passed state is the default
-        suite.test_cases.append(case)
-
-    return suite
+            return str(report.step.plan.execute.read(log))
+        except AttributeError:
+            return ''
+
+    environment = default_template_environment()
+
+    # Use a FileSystemLoader for a non-custom flavor
+    if flavor != CUSTOM_FLAVOR_NAME:
+        environment.loader = FileSystemLoader(searchpath=Path('tmt/steps/report/junit/templates/'))
+        template_path = Path(tmt.utils.resource_files(
+            Path(f'steps/report/junit/templates/{flavor}.xml.j2')))
+
+    environment.filters.update({
+        'read_log': _read_log,
+        'duration_to_seconds': duration_to_seconds,
+        })
+
+    # Explicitly enable the autoescape because it's disabled by default by TMT
+    # (see /teemtee/tmt/issues/2873 for more info.
+    environment.autoescape = select_autoescape(enabled_extensions=('xml'))
+
+    xml_data = render_template_file(
+        template_path,
+        environment,
+        RESULTS=results_context,
+        PLAN=report.step.plan,
+        INCLUDE_OUTPUT_LOG=report.data.include_output_log,
+        **extra_variables)
+
+    # Try to use lxml to check the flavor XML schema and prettify the final XML
+    # output.
```

Comment:
FTR, I for one would not bother with prettifying the output - it's been rendered by a template whose owner had full and absolute control over indentation and line splitting - if they decided that their elements shall be indented by 13 tabs for every odd level, then they should have such indentation. Their template, their rules, who are we to mess up their team's perfectly selected formatting rules :) Validation, absolutely, but I would advise against reformatting and changing the output in any way.

Reply:
I agree that we should probably disable the pretty printing for the `custom` flavor/template. However, I still believe it should be enabled for predefined TMT JUnit flavors. Without it, the XML output isn't very clean - jinja macros add a lot of whitespace and poor indentation.

I spent some time modifying the templates (playing with [Jinja whitespace control](https://jinja.palletsprojects.com/en/3.0.x/templates/#whitespace-control)) to improve the output, but this made the Jinja templates much harder to read.

This is an example output with `pretty_print=False` and `remove_blank_text=False`:
```
<?xml version='1.0' encoding='utf-8'?>
<testsuites disabled="0" errors="1" failures="1" tests="7" time="1">
    <testsuite name="/plans/example&lt;br&gt;" disabled="0" errors="1" failures="1" skipped="1" tests="7" time="1">

                <testcase name="/tests/call-dmesg&quot;&lt;br&gt;">

                        <system-out>test1
</system-out>
                </testcase>

                <testcase name="/tests/mine-rep-multi">
                        <failure type="failure" message="fail"/>

                </testcase>

                <testcase name="/tests/mine-rep-pass">

                </testcase>

                <testcase name="/tests/please-fail">
                        <error type="error" message="error"/>

                </testcase>

                <testcase name="/tests/please-skip">
                        <skipped type="skipped" message="info">skipped
</skipped>

                        <system-out>skipped
</system-out>
                </testcase>

                <testcase name="/tests/sleep" time="1">

                </testcase>

                <testcase name="/tests/special-output">

                        <system-out>&lt;h2&gt;ahoj"jak"se&amp;mas&lt;/h2&gt;
</system-out>
                </testcase>
    </testsuite>

</testsuites>
```

Reply:
Or, we can simply provide `--prettify` or `--pretty` option to enable this behavior, and let the user decide. Can be enabled by default for non-`custom` flavors, disabled by default for `custom` ones, and everyone would be happy :)

Reply:
The custom template no longer even needs to be in XML format:

Example test input: https://github.com/teemtee/tmt/blob/7148b056697843c0707970b545c84004af7572d8/tests/report/junit/data/custom.xml.j2

Reply:
Yeah, that's a correct observation. It's similar to `export` where you can go with `--how=template`, and use whatever Jinja template you wish, or `--how=rst` which takes its own Jinja template and calls `template` plugin to render it.

Maybe we could do the same here - probably not here and now, but it would make sense to have a `report/template` plugin that would focus on rendering results, and change `junit` plugin to 1. ship templates for various flavors, 2. call `template` plugin, 3. perform validation & pretty-printing, as bells & whistles on top of the actual junit production. I wouldn't let `junit` render anything for now, it should be a XML outcome, otherwise, we risk `junit` would start rendering Asciidoc representation of results because users realize they can use any template - but we could give them `report/template` someday in the future.

Reply:
Ok, I agree. I'll change the code to disallow non-XML custom templates, just to be sure users will not misuse it. I also added info about it to the description of this PR.